src/analysis/refactor_extract_region.ml

```diff
@@ -0,0 +1,479 @@
+  |> Msource.make
+
+type analysis = { bounded_vars : Path.t list; binding_kind : rec_flag }
+
+and rec_flag = Non_recursive | Rec_and
+
+type extraction =
+  { expr : Typedtree.expression;  (** Expression that being extracted *)
+    expr_env : Env.t;  (** Environment of the extracted expression *)
+    toplevel_item : toplevel_item;
+        (** The value binding toplevel or class declaration enclosing the extracted expression. *)
+    name : extraction_name;  (** Binding name of the extracted expression. *)
+    gen_binding_kind : rec_flag;
+    generated_binding : generated_binding;
+    generated_call : generated_call;
+    call_need_parenthesis : bool
+        (** Sometime we must parenthised call in order to type check. *)
+  }
+
+and extraction_name = Default of { basename : string } | Fixed of string
+
+and toplevel_item =
+  { rec_flag : Asttypes.rec_flag;
+    env : Env.t;
+    loc : Location.t;
+    kind : toplevel_item_kind
+  }
+(* A convenient type for grouping info. *)
+
+and toplevel_item_kind = Let of Typedtree.value_binding list | Class_decl
+
+and generated_binding =
+  name:string -> body:Parsetree.expression -> Parsetree.structure_item
+
+and generated_call = name:string -> Parsetree.expression
+
+let is_recursive = function
+  | { rec_flag = Asttypes.Recursive; _ } -> true
+  | { rec_flag = Nonrecursive; _ } -> false
+
+let rec find_pattern_var : type a. a Typedtree.general_pattern -> Path.t list =
+ fun { Typedtree.pat_desc; _ } ->
+  match pat_desc with
+  | Typedtree.Tpat_var (ident, _, _) -> [ Pident ident ]
+  | Tpat_tuple pats -> List.concat_map ~f:find_pattern_var pats
+  | Tpat_alias (pat, ident, _, _) -> Pident ident :: find_pattern_var pat
+  | Tpat_construct (_, _, pats, _) -> List.concat_map ~f:find_pattern_var pats
+  | Tpat_variant (_, Some pat, _) -> find_pattern_var pat
+  | Tpat_record (fields, _) ->
+    List.concat_map ~f:(fun (_, _, field) -> find_pattern_var field) fields
+  | Tpat_array arr -> List.concat_map ~f:find_pattern_var arr
+  | Tpat_lazy pat | Tpat_exception pat -> find_pattern_var pat
+  | Tpat_value pat ->
+    find_pattern_var (pat :> Typedtree.value Typedtree.general_pattern)
+  | Tpat_or (l, r, _) -> find_pattern_var l @ find_pattern_var r
+  | _ -> []
+
+let rec occuring_vars_path node =
+  let loop acc node =
+    match node.Browse_tree.t_node with
+    | Browse_raw.Expression { exp_desc = Texp_ident (path, _, _); _ } ->
+      path :: acc
+    | Pattern pat -> find_pattern_var pat @ acc
+    | _ ->
+      Lazy.force node.t_children
+      |> List.concat_map ~f:occuring_vars_path
+      |> List.append acc
+  in
+  loop [] node |> Path.Set.of_list |> Path.Set.elements |> List.rev
+  |> List.filter ~f:(fun path ->
+         (* TODO: fix this *)
+         Ident.name (Path.head path) <> "Stdlib")
+
+let analyze_expr expr expr_env ~toplevel_item =
+  let is_value_unbound path =
+    let is_bound path env =
+      try
+        let _ = Env.find_value path env in
```

Comment:
`Env.bound_vale` would be a better choice here.

Reply:
I use `Env.find_value` instead of `Env.bound_value` because the latter takes a string instead of a path.